File: ios_app/QueenMamaIOS/Services/TranscriptBuffer.swift
//
//  TranscriptBuffer.swift
//  QueenMama
//
//  Accumulates transcript updates and flushes in batches to reduce UI re-renders
//  and SwiftData writes. Instead of updating on every word, batches updates every 500ms.
//

import Foundation

@MainActor
final class TranscriptBuffer: ObservableObject {
    // MARK: - Configuration

    /// Time to wait before flushing accumulated transcript
    private let flushInterval: TimeInterval = 0.5

    // MARK: - State

    private var pendingText = ""
    private var flushTimer: Timer?
    private var totalUpdates: Int = 0
    private var totalFlushes: Int = 0

    // MARK: - Callbacks

    /// Called when buffered transcript is ready (batched update)
    var onFlush: ((String) -> Void)?

    // MARK: - Initialization

    init() {}

    // MARK: - Public Methods

    /// Append transcript text to the buffer
    /// Will be flushed after flushInterval or when flush() is called
    func append(_ text: String) {
        pendingText += text + " "
        totalUpdates += 1

        // Schedule flush if not already pending
        scheduleFlush()
    }

    /// Force flush any pending transcript
    func flush() {
        flushTimer?.invalidate()
        flushTimer = nil

        guard !pendingText.isEmpty else { return }

        let textToFlush = pendingText
        pendingText = ""
        totalFlushes += 1

        // Log periodically
        if totalFlushes % 20 == 0 {
            let batchRatio = totalFlushes > 0 ? totalUpdates / totalFlushes : 0
            print("[TranscriptBuffer] Stats: \(totalUpdates) updates batched into \(totalFlushes) flushes (avg \(batchRatio) updates/batch)")
        }

        onFlush?(textToFlush)
    }

    /// Reset the buffer (call when stopping session)
    func reset() {
        // Flush any remaining text first
        flush()

        pendingText = ""
        flushTimer?.invalidate()
        flushTimer = nil

        print("[TranscriptBuffer] Reset - \(totalUpdates) updates batched into \(totalFlushes) flushes")
        totalUpdates = 0
        totalFlushes = 0
    }

    /// Start the buffer (call when starting session)
    func start() {
        pendingText = ""
        totalUpdates = 0
        totalFlushes = 0
        print("[TranscriptBuffer] Started with \(flushInterval)s flush interval")
    }

    // MARK: - Private Methods

    private func scheduleFlush() {
        // Don't reschedule if timer is already running
        guard flushTimer == nil else { return }

        flushTimer = Timer.scheduledTimer(withTimeInterval: flushInterval, repeats: false) { [weak self] _ in
            Task { @MainActor in
                self?.flushTimer = nil
                self?.flush()
            }
        }
    }
}

File: win_app_2/src/services/transcription/transcriptBuffer.ts
/**
 * TranscriptBuffer - Batches transcript updates to reduce UI redraws
 *
 * Accumulates transcript text and flushes every 500ms,
 * reducing SwiftUI-equivalent re-renders by ~3-5x.
 * Matches macOS TranscriptBuffer behavior.
 */

import { createLogger } from '@/lib/logger'

const log = createLogger('TranscriptBuffer')

const FLUSH_INTERVAL = 500 // ms - matches macOS

export class TranscriptBuffer {
  private pendingText = ''
  private flushTimer: ReturnType<typeof setInterval> | null = null
  private onFlush: ((text: string) => void) | null = null
  private totalUpdates = 0
  private totalFlushes = 0

  start(callback: (text: string) => void): void {
    this.onFlush = callback
    this.pendingText = ''
    this.totalUpdates = 0
    this.totalFlushes = 0

    this.flushTimer = setInterval(() => {
      this.flush()
    }, FLUSH_INTERVAL)
  }

  stop(): void {
    // Force flush any remaining text
    this.flush()

    if (this.flushTimer) {
      clearInterval(this.flushTimer)
      this.flushTimer = null
    }

    this.onFlush = null

    if (this.totalFlushes > 0) {
      const ratio = (this.totalUpdates / this.totalFlushes).toFixed(1)
      log.info(`Session stats: ${this.totalUpdates} updates → ${this.totalFlushes} flushes (${ratio}x reduction)`)
    }
  }

  append(text: string): void {
    this.totalUpdates++
    const separator = this.pendingText.length > 0 ? ' ' : ''
    this.pendingText += separator + text
  }

  private flush(): void {
    if (this.pendingText.length === 0) return

    this.totalFlushes++
    const text = this.pendingText
    this.pendingText = ''
    this.onFlush?.(text)

    // Log batch ratio periodically
    if (this.totalFlushes % 20 === 0) {
      const ratio = (this.totalUpdates / this.totalFlushes).toFixed(1)
      log.info(`Batch ratio: ${ratio}x (${this.totalUpdates} updates → ${this.totalFlushes} flushes)`)
    }
  }
}

// Singleton instance
export const transcriptBuffer = new TranscriptBuffer()
